devito/types/sparse.py

```diff
@@ -1082,92 +937,17 @@ def __init_finalize__(self, *args, **kwargs):
-        self.r = r
-
-        gridpoints = SubFunction(name="%s_gridpoints" % self.name, dtype=np.int32,
-                                 dimensions=(self.indices[-1], Dimension(name='d')),
-                                 shape=(self.npoint, self.grid.dim), space_order=0,
-                                 parent=self)
-
-        gridpoints_data = kwargs.get('gridpoints', None)
-        assert(gridpoints_data is not None)
-        gridpoints.data[:] = gridpoints_data[:]
-        self._gridpoints = gridpoints
-
-        interpolation_coeffs = SubFunction(name="%s_interpolation_coeffs" % self.name,
-                                           dimensions=(self.indices[-1],
-                                                       Dimension(name='d'),
-                                                       Dimension(name='i')),
-                                           shape=(self.npoint, self.grid.dim,
-                                                  self.r),
-                                           dtype=self.dtype, space_order=0,
-                                           parent=self)
-        coefficients_data = kwargs.get('interpolation_coeffs', None)
-        assert(coefficients_data is not None)
-        interpolation_coeffs.data[:] = coefficients_data[:]
-        self._interpolation_coeffs = interpolation_coeffs
-        warning("Ensure that the provided interpolation coefficient and grid point " +
-                "values are computed on the final grid that will be used for other " +
-                "computations.")
-
-    def interpolate(self, expr, offset=0, increment=False, self_subs={}):
-        """
-        Generate equations interpolating an arbitrary expression into ``self``.
+        gridpoints = kwargs.get('gridpoints')
+        interpolation_coeffs = kwargs.get('interpolation_coeffs')
 
-        Parameters
-        ----------
-        expr : expr-like
-            Input expression to interpolate.
-        offset : int, optional
-            Additional offset from the boundary.
-        increment: bool, optional
-            If True, generate increments (Inc) rather than assignments (Eq).
-        """
-        expr = indexify(expr)
-
-        p, _, _ = self.interpolation_coeffs.indices
-        dim_subs = []
-        coeffs = []
-        for i, d in enumerate(self.grid.dimensions):
-            rd = DefaultDimension(name="r%s" % d.name, default_value=self.r)
-            dim_subs.append((d, INT(rd + self.gridpoints[p, i])))
-            coeffs.append(self.interpolation_coeffs[p, i, rd])
-        # Apply optional time symbol substitutions to lhs of assignment
-        lhs = self.subs(self_subs)
-        rhs = prod(coeffs) * expr.subs(dim_subs)
-
-        return [Eq(lhs, lhs + rhs)]
-
-    def inject(self, field, expr, offset=0):
-        """
-        Generate equations injecting an arbitrary expression into a field.
+        self.interpolator = PrecomputedInterpolator(self, r, gridpoints,
```

Comment:
I think if we do as suggested above, this line becomes unnecessary as executed automatically through `super(...)` while implicitly using the right interpolator type

Reply:
As discussed above

Reply:
I think github isn't showing me this "as discussed above"? could you paste it or write it again please?

Reply:
I am referencing this comment:
> Can be kept as a TODO because the constructors for LinearInterpolator and PrecomputedInterpolator take different arguments for now. Not easy to homogenise yet.